### PR TITLE
Initial yocto git based recipe

### PR DESCRIPTION
**Pull request by eli-zr**
Added a base recipe that fetches latest version from git

### CODE REVIEW

File: yocto/myencfs_git.bb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+SUMMARY = "myencfs example application"
+DESCRIPTION="A simple directory encryption/decryption using authenticated encryption"
+LICENSE = "BSD"
+LIC_FILES_CHKSUM = "file://COPYING;md5=9859f02baa869af387514f9e5dd8cc3e \
+                    file://debian/copyright;md5=358aa3ad2f63e0c87d59d3150ef3fd7e"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: This is debian license, not the package license, is it custom to add it? It does not feel "right" :)

File: yocto/myencfs_git.bb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+SRC_URI = "git://github.com/alonbl/myencfs.git;protocol=https"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: interesting, I did not expect this format `protocol=https` but an `https://` URL.

File: yocto/myencfs_git.bb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+SUMMARY = "myencfs example application"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: Is this `_git` suffix of package will be treated in a special way? what if we have in a directory both `myencfs_0.0.0.bb` and `myencfs_git.bb`? This is the reason in Gentoo we use 9999 version for live ebuilds.

File: yocto/myencfs_git.bb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+MYENCFS_CRYPTO ?= "openssl_crypto"
+# List of enabled features that can be overriden
+PACKAGECONFIG ?= "tools encrypt decrypt bio-file test fuse static-libs ${MYENCFS_CRYPTO}"
+
+
+PACKAGECONFIG[tools]	= "--enable-tool"
+PACKAGECONFIG[encrypt]	= "--enable-encrypt"
+PACKAGECONFIG[decrypt]	= "--enable-decrypt"
+PACKAGECONFIG[bio-file]	= "--enable-bio-file"
+PACKAGECONFIG[test]	= "--enable-tests"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: as we never build for native, we can avoid enabling tests, so this can be removed.

File: yocto/myencfs_git.bb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+PACKAGECONFIG[static-libs]	= "--enable-all-static"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: In Gentoo this enables `--enable-static` not the extreme `all-static`, was there a reason to apply all?

File: yocto/myencfs_git.bb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+PACKAGECONFIG[openssl_crypto]	= " \
+				--with-crypto=openssl, \
+				, \
+				openssl, \
+				openssl, \
+				, \
+				mbedtls_crypto \
+				"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: I only see documentation for 4 parameters, what is the meaning of mbedtls_crypto  in this context? can you please add support for wolfssl as well so we can see the entire picture, in Gentoo I did not have wolf so I could not add it.

File: yocto/myencfs_git.bb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+LIC_FILES_CHKSUM = "file://COPYING;md5=9859f02baa869af387514f9e5dd8cc3e \
+                    file://debian/copyright;md5=358aa3ad2f63e0c87d59d3150ef3fd7e"
+
+SRC_URI = "git://github.com/alonbl/myencfs.git;protocol=https"
+
+SRCREV = "${AUTOREV}"
+S = "${WORKDIR}/git"
+
+# Crypto can be either openssl_crypto or mbedtls_crypto
+MYENCFS_CRYPTO ?= "openssl_crypto"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: I prefer `crypto_*` if possible.

File: yocto/myencfs_git.bb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+# List of enabled features that can be overriden
+PACKAGECONFIG ?= "tools encrypt decrypt bio-file test fuse static-libs ${MYENCFS_CRYPTO}"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: In the examples I can see double '?' what is the meaning?

File: yocto/myencfs_git.bb [hunk 9]
```diff
@@ -0,0 +1,46 @@
+SUMMARY = "myencfs example application"
+DESCRIPTION="A simple directory encryption/decryption using authenticated encryption"
+LICENSE = "BSD"
+LIC_FILES_CHKSUM = "file://COPYING;md5=9859f02baa869af387514f9e5dd8cc3e \
+                    file://debian/copyright;md5=358aa3ad2f63e0c87d59d3150ef3fd7e"
+
+SRC_URI = "git://github.com/alonbl/myencfs.git;protocol=https"
+
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: don't you need to inherit autotools? does it automatically run autoreconf?

File: yocto/myencfs_git.bb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+SRCREV = "${AUTOREV}"
+S = "${WORKDIR}/git"
+
+# Crypto can be either openssl_crypto or mbedtls_crypto
+MYENCFS_CRYPTO ?= "openssl_crypto"
+# List of enabled features that can be overriden
+PACKAGECONFIG ?= "tools encrypt decrypt bio-file test fuse static-libs ${MYENCFS_CRYPTO}"
+
+
+PACKAGECONFIG[tools]	= "--enable-tool"
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: usually it is best practice to provide both enable and disable as you never knows what the default of the package and when it changes. Gentoo is doing that automatically, I see that yocto does not have these helpers.

File: yocto/myencfs_git.bb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+PACKAGECONFIG[mbedtls_crypto] = " \
+				--with-crypto=mbed, \
+				, \
+				mbedtls, \
+				mbedtls, \
+				, \
+				openssl_crypto \
+				"
+
+PACKAGECONFIG[fuse] = "--enable-fuse --enable-decrypt --enable-bio-file, \
```
Comment (alonbl) [inline yocto/myencfs_git.bb]: how do we produce an error if both openssl and mbed were specified?

**Review by alonbl — COMMENTED** [review]
Thanks!!!!

I have lots of questions :)

File: yocto/myencfs_git.bb (hunk 1 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: it should be the package license which is BSD 3-Clause and not BSD as I specified. Would fix that.

File: yocto/myencfs_git.bb (hunk 2 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Under Yocto, fetching from git should use git:// prefix, and the protocol may vary (SSH, https etc.).

File: yocto/myencfs_git.bb (hunk 3 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: I saw your convention of 9999 as the PV and I didn't ever see something similar under Yocto. Typically when you fetch the latest and greatest you just put git as the PV but not really use it in the recipe. Under Yocto I would put most of the recipe logic in an include file, and create two recipes, one for git and one for fetching a tar package. In each recipe I would just set SRC_URI and include the header. This is a more common practice under Yocto, at least from my experience.

File: yocto/myencfs_git.bb (hunk 1 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Sorry, I didn't provide the complete answer. I use a tool to generate the skeleton for a package and it was placed there for me and I'm not sure why is it needed. Let me review that again but I believe I can omit that.

File: yocto/myencfs_git.bb (hunk 4 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Actually the enabling of the features is done in this line:
PACKAGECONFIG ?= "tools encrypt decrypt bio-file test fuse static-libs ${MYENCFS_CRYPTO}"

I would remove test from the default settings.

File: yocto/myencfs_git.bb (hunk 7 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Sure, will change that

File: yocto/myencfs_git.bb (hunk 4 above)
Comment (alonbl) [reply yocto/myencfs_git.bb]: No, I mean that there is no reason to keep the test feature, it will never work under yocto cross compile

**Comment by alonbl** [comment]
Don't we also need `layer.conf`? I see many examples such as [1].

[1] https://github.com/wolfSSL/meta-wolfssl/blob/master/conf/layer.conf

File: yocto/myencfs_git.bb (hunk 9 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Definitely, my bad. I worked on several stations and not sure how I dropped that when moving this file

File: yocto/myencfs_git.bb (hunk 5 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: So it was my wrong interpretation. Will change that as well.

File: yocto/myencfs_git.bb (hunk 10 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Actually it is defined implicitly. PACKAGECONFIG has six parameters which can be explicitly defined. I used it in the crypto_openssl for example. The second empty field is the disable field which is left empty. When I don't set --enable-tool for example, it is the disablement of tools. I hope I was clear

File: yocto/myencfs_git.bb (hunk 11 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: This is the magic of PACKAGECONFIG. The sisxth param is the "confilcts with" parameter. See for example crypto_openssl conflicts with crypto_mbedtls. If both would be enabled, the build will faill with a specific message.

**Comment by eli-zr** [comment]
I didn't yet add a layer. only a recipe so you don't see layer.conf.
Generally you don't create a layer for a single  simple. I can add one though for the sake of the example

**Comment by alonbl** [comment]
> I didn't yet add a layer. only a recipe so you don't see layer.conf. Generally you don't create a layer for a single simple. I can add one though for the sake of the example

yes, for the sake of example as I've done for the Gentoo, I think it will be great.

File: yocto/myencfs_git.bb (hunk 10 above)
Comment (alonbl) [reply yocto/myencfs_git.bb]: I think you should explicitly disable the feature as if the upstream modifies the defaults, you may end up with unintentionally enabled feature.

what are the 5th and 6th parameters? where can I read?

File: yocto/myencfs_git.bb (hunk 11 above)
Comment (alonbl) [reply yocto/myencfs_git.bb]: oh... how do you conflict with multiple? where can I read about it? I can find only this[1] source.

[1] https://www.yoctoproject.org/docs/2.4.2/ref-manual/ref-manual.html#var-PACKAGECONFIG

File: yocto/myencfs_git.bb (hunk 6 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: See the following documentation. It explains the 6 params of PACKAGECONFIG
https://docs.yoctoproject.org/3.2.3/ref-manual/ref-variables.html#term-PACKAGECONFIG

File: yocto/myencfs_git.bb (hunk 10 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: I've provided you with a link in a previous comment

File: yocto/myencfs_git.bb (hunk 11 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: According to documentation you can specify multiple conflicts as you can specify multiple depends or rdepends. I'll test that when I add wolfssl. As you probably saw PACKAGECONFIG functionality has been upgraded over the yocto releases, Initially, as you specified it has only supported 4 params.

File: yocto/myencfs_git.bb (hunk 10 above)
Comment (eli-zr) [reply yocto/myencfs_git.bb]: Added an explicit disable

**Comment by eli-zr** [comment]
Layer is included

**Comment by alonbl** [comment]
Thanks!

Next time please squash patches and force push so that it will be a single perfect patch.

I added Makefile.am so it will be added to dist.
I added EXTRA_OECONF so that we have a build-id.
I also reformatted the whole thing to be more consistent.
Fixed some typo (single dash for bio enable).

Questions:
1. Why do you use single `?=` while I see in many example double `??=`
2. Do I understand correctly that in `layer.conf` you can set `BBFILE_PATTERN` so that it will work also when the myencfs git will be checked out directly? I mean put the prefix of `yocto/meta-myencfs`?

**Comment by alonbl** [comment]
I also added bit of documentation and squashed it all.

Comment is [here](https://github.com/alonbl/myencfs/commit/43a78478e771e910615c49e0598f1b96b37f8a1c)

**Comment by eli-zr** [comment]
1. Both ?= & ??= are setting a default value. The difference is that with ?= the assignment of the default value is done immediately while with ??= the assignment is done after bitbake completes parsing the metadata. It is a matter of your requirement but still it is possible to override the default value.
2. I'm not sure I understand the question. Following is the documentation of this variable:
BBFILE_PATTERN    Variable that expands to match files from [BBFILES](https://www.yoctoproject.org/docs/3.1.3/ref-manual/ref-manual.html#var- BBFILES) in a particular layer. This variable is used in the conf/layer.conf file and must be suffixed with the name of the specific layer

**Comment by alonbl** [comment]
I just wounder why all the examples and in core layer they are using `??=` while we `?=`...

For the `layer.conf` I ask what happens if I checkout the myencfs git repository directly as a layer, will bitbake will be able to use it directly?

For example in Gentoo there is no way to checkout the `https://github.com/alonbl/myencfs` git repos and use it as a repository as it must seek the `metadata/layout.conf` at the root, if not found it is not a repository. They should have added a way to look into a subdirectory...

**Pull request closed without merging**